### PR TITLE
plugins/barbecue: init + tests

### PR DESCRIPTION
**Pull request by hmajid2301**
PR to add support for the [barbecue-nvim](https://github.com/utilyre/barbecue.nvim plugin)

### CODE REVIEW

File: plugins/bufferlines/barbecue.nix [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+      showBasename = helpers.defaultNullOpts.mkBool true ''
+        Whether to display file name.
+      '';
+
+      showModified = helpers.defaultNullOpts.mkBool false ''
+        Whether to replace file icon with the modified symbol when buffer is modified.
+      '';
+
+      modified = helpers.defaultNullOpts.mkStr ''function(bufnr) return vim.bo[bufnr].modified end,'' ''
```
Comment (hmajid2301) [inline plugins/bufferlines/barbecue.nix:61]: Not sure the best type for this lua function, I saw some other packages doing this!

Comment (hmajid2301) [reply plugins/bufferlines/barbecue.nix:61]: I guess this is also whats throwing an error in CI (with the `lead_custom_section`. Thanks.

Comment (GaetanLepage) [reply plugins/bufferlines/barbecue.nix:61]: ->Question: When to choose between an `types.str` and a `helpers.rawType` option for lua code ?

- If the upstream option expects only lua code, we use `types.str`. There is no ambiguity and the user doesn't have to bother with providing a `rawType` value.
In this case, we do the conversion ourselves in the `config` section of the plugin: `foo = helpers.ifNonNull' foo (helpers.mkRaw foo);`
- If the upstream option accept both strings and lua code (or when there is some kind of ambiguity), we rely on `helpers.rawType`. In this case, there is no need to convert the provided value.

File: plugins/bufferlines/barbecue.nix [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+      attachNavic = helpers.defaultNullOpts.mkBool true ''
+        Whether to attach navic to language servers automatically.
+      '';
+
+      createAutocmd = helpers.defaultNullOpts.mkBool true ''
+        Whether to create winbar updater autocmd.
+      '';
+
+      includeBuftypes = mkListStr ''[ "" ]'' ''
```
Comment (GaetanLepage) [inline plugins/bufferlines/barbecue.nix:27]:
```suggestion
      includeBuftypes = mkListStr ''[""]'' ''
```

File: plugins/bufferlines/barbecue.nix [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        Buftypes to enable winbar in.
+      '';
+
+      excludeFiletypes = mkListStr ''[ "netrw" "toggleterm" ]'' ''
```
Comment (GaetanLepage) [inline plugins/bufferlines/barbecue.nix:31]:
```suggestion
      excludeFiletypes = mkListStr ''["netrw" "toggleterm"]'' ''
```

File: plugins/bufferlines/barbecue.nix [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+      showNavic = helpers.defaultNullOpts.mkBool true ''
+        Whether to show/use navic in the winbar.
+      '';
+
+      leadCustomSection = helpers.defaultNullOpts.mkStr ''function() return " " end,'' ''
+            Get leading custom section contents.
+
+        NOTE: This function shouldn't do any expensive actions as it is run on each render.
+      '';
```
Comment (GaetanLepage) [inline plugins/bufferlines/barbecue.nix:75]:
```suggestion
      leadCustomSection =
        helpers.defaultNullOpts.mkStr
        ''
          function()
            return " "
          end
        ''
        ''
          Get leading custom section contents.
          NOTE: This function shouldn't do any expensive actions as it is run on each render.
        '';
```
Prefer this formatting for those options.

File: tests/test-sources/plugins/bufferlines/barbecue.nix [hunk 5]
```diff
@@ -0,0 +1,61 @@
+{
+  empty = {
+    plugins.barbecue.enable = true;
+  };
+
+  defaults = {
+    plugins.barbecue = {
+      enable = true;
+      attachNavic = true;
```
Comment (GaetanLepage) [inline tests/test-sources/plugins/bufferlines/barbecue.nix:9]:
```suggestion
      enable = true;

      attachNavic = true;
```

**Review by GaetanLepage — COMMENTED** [review]
Very good looking !
Here are some first minor remarks.

File: plugins/bufferlines/barbecue.nix (hunk 1 above)
Comment (hmajid2301) [reply plugins/bufferlines/barbecue.nix:61]: Okay just to replay it back to you; we carry on using the str type. I just need to update the config section with this helpers function (?) you defined.

File: plugins/bufferlines/barbecue.nix (hunk 2 above)
Comment (hmajid2301) [reply plugins/bufferlines/barbecue.nix:27]: sure!

File: plugins/bufferlines/barbecue.nix (hunk 4 above)
Comment (hmajid2301) [reply plugins/bufferlines/barbecue.nix:75]: yeh looks much better, will update the others.

File: plugins/bufferlines/barbecue.nix [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+        exclude_filetypes = excludeFiletypes;
+        show_dirname = showDirname;
+        show_basename = showBasename;
+        show_modified = showModified;
+        show_navic = showNavic;
+        lead_custom_section = leadCustomSection;
+        custom_section = customSection;
+        context_follow_icon_color = contextFollowIconColor;
+      }
+      // cfg.extraOptions;
```
Comment (GaetanLepage) [inline plugins/bufferlines/barbecue.nix:181]:
```suggestion
        attach_navic = attachNavic;
        create_autocmd = createAutocmd;
        include_buftypes = includeBuftypes;
        exclude_filetypes = excludeFiletypes;
        modifiers = {
          inherit (modifiers)
            dirname
            basename
            ;
        };
        show_dirname = showDirname;
        show_basename = showBasename;
        show_modified = showModified;
        modified = helpers.ifNonNull' modified; (helpers.mkRaw modified);
        show_navic = showNavic;
        lead_custom_section = helpers.ifNonNull' leadCustomSection (helpers.mkRaw leadCustomSection);
        custom_section = helpers.ifNonNull' customSection (helpers.mkRaw customSection);
        inherit theme;
        context_follow_icon_color = contextFollowIconColor;
        symbols = {
          inherit (symbols);
            modified
            ellipsis
            separator
            ;
        };
        inherit kinds;       
      }
      // cfg.extraOptions;
```

Comment (GaetanLepage) [reply plugins/bufferlines/barbecue.nix:181]: 1. You had forgotten to convert (`helpers.mkRaw`) the values of the options that expect lua code
2. Keep the same order in `setupOptions` as the one used when declaring options.

Comment (hmajid2301) [reply plugins/bufferlines/barbecue.nix:181]: 1. yeh I was waiting to confirm above if 1 was the right approach!
2. ahh yes the ordering good shout

**Review by GaetanLepage — APPROVED** [review]
LGTM

**Pull request merged**